dali/python/nvidia/dali/pipeline.py

```diff
@@ -990,3 +996,63 @@ def iter_setup(self):
         For example, one can use this function to feed the input
         data from NumPy arrays."""
         pass
+
+
+def _discriminate_args(**kwargs):
+    """Split args on those applicable to Pipeline constructor and the rest."""
+    fca = inspect.getfullargspec(Pipeline.__init__)  # Full Ctor Args
+    ctor_args = dict(filter(lambda x: x[0] in fca.args or x[0] in fca.kwonlyargs, kwargs.items()))
+    fn_args = dict(filter(lambda x: x[0] not in ctor_args.keys(), kwargs.items()))
+    return ctor_args, fn_args
+
+
+def pipeline(fn=None, **pipeline_kwargs):
+    """
+    Decorator for wrapping functions that define DALI pipelines.
+
+    **Usage**
+
+    First, implement a function, that defines a DALI pipeline.
+    Such function shall return DALI's DataNodes. These return
+    values will denote output from the pipeline. You can
+    decorate this function with ``@pipeline``::
+
+        @nvidia.dali.pipeline.pipeline
+        def pipe(flip_vertical, flip_horizontal):
+            data, _ = fn.file_reader(file_root=images_dir)
+            img = fn.image_decoder(data, device="mixed")
+            flipped = fn.flip(img, horizontal=flip_horizontal, vertical=flip_vertical)
+            return flipped, img
+
+    When creating a Pipeline object using the decorated function,
+    you can pass any number of ``Pipeline.__init__`` arguments as
+    keyword-args to the call::
+
+        my_pipe = pipe(0, batch_size=32, num_threads=1, device_id=0, flip_horizontal=1)
+        my_pipe.build()
+
+    Additionally, decorator can accept Pipeline constructor
+    parameters. Please note, that in this case you must use
+    keyword args only. Any Pipeline constructor parameter
+    passed later at pipeline instantiation will overwrite
+    the decorator-defined params::
+
+        @nvidia.dali.pipeline.pipeline(batch_size=32, num_threads=3)
+        def pipe():
+            data = fn.external_source(source=my_generator)
+            return data
+
+        my_pipe = pipe(batch_size=128)  # batch_size=128 overwrites batch_size=32
```

Comment:
I'm also thinking if the doc should mention that this is equivalent to using `with pipe:` and set_outputs?

Reply:
Done